File: kv_store/delete_key.py
# delete_key.py
# from kv_store import kv_store
from trie_kv_store import kv_store
from lock_manager import kv_lock
import threading

# Helper function to handle retries for deleting a key
def handle_delete_thread(key, timeout=0.01):
    result = []
    # Call thread, append results to result
    thread = threading.Thread(target=lambda: result.append(delete_key(key)))
    thread.start()
    thread.join(timeout)
    
    # If timeout occurs, send another thread out
    if thread.is_alive():
        thread = threading.Thread(target=lambda: result.append(delete_key(key)))
        thread.start()
        thread.join(timeout)
        
    # Both threads failed - return failure
    if thread.is_alive():
        return None
    return result


def delete_key(key):
    # Safely delete the key from kv_store using lock
    with kv_lock:
        res = kv_store.delete(key)
        if res is not None:
            return 0  # Key successfully deleted
        return -1  # Key not found

File: Rust/src/delete_key.rs
// delete_key.rs
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;
use crate::kv_store::KVStore;

pub fn handle_delete_thread(key: String, kv_store: Arc<Mutex<KVStore>>, timeout: f64) -> Option<i32> {
    let key_clone = key.clone();
    let store_clone = Arc::clone(&kv_store);
    
    // First attempt
    let handle = thread::spawn(move || {
        delete_key(key_clone, store_clone)
    });

    match handle.join_timeout(Duration::from_secs_f64(timeout)) {
        Ok(result) => return Some(result),
        Err(_) => {
            // Second attempt
            let key_clone = key.clone();
            let store_clone = Arc::clone(&kv_store);
            let handle = thread::spawn(move || {
                delete_key(key_clone, store_clone)
            });

            match handle.join_timeout(Duration::from_secs_f64(timeout)) {
                Ok(result) => Some(result),
                Err(_) => None,
            }
        }
    }
}

pub fn delete_key(key: String, kv_store: Arc<Mutex<KVStore>>) -> i32 {
    // Safely delete the key from kv_store using lock
    let mut store = kv_store.lock().unwrap();
    match store.delete(&key) {
        Some(_) => {
            println!("delete res: success");
            0  // Key successfully deleted
        },
        None => {
            println!("delete res: key not found");
            -1  // Key not found
        }
    }
}
